### ISSUE TITLE
Multi-file grammars break repeated langium:generate calls

### ISSUE THREAD

**Issue by snarkipus** [langium/langium#882]
See Discussion #879 

Langium version: 1.0.1
Package name: langium

## Steps To Reproduce

1. Create multi-file grammar with internal import cross-dependencies (see code example)
2. Run `langium:watch` 
3. Save/change the _super grammar_ file: `block-madness.langium` to trigger generation
4. First attempt should fail (see below)
3. Save/change the _super grammar_ file: `block-madness.langium` to trigger a 2nd generation
4. Second attempt should succeed

Link to code example:
[Fresh Grammar - Minimal Example Repo](https://github.com/snarkipus/langium-bug)
note: apologize that I accidentally omitted  `.gitignore`, so it has all the build artifacts

## The current behavior
```console
[10:30:13] File change detected. Starting compilation...
Attempted reference resolution before document reached ComputedScopes state (file:///home/snarkipus/Projects/langium/langium-race/block-madness/src/language-server/block-madness-A.langium).
Attempted reference resolution before document reached ComputedScopes state (file:///home/snarkipus/Projects/langium/langium-race/block-madness/src/language-server/block-madness-B.langium).
Attempted reference resolution before document reached ComputedScopes state (file:///home/snarkipus/Projects/langium/langium-race/block-madness/src/language-server/block-madness-C.langium).
[10:30:13] src/language-server/block-madness.langium:9:15 - Could not resolve reference to AbstractRule named 'Person'.
[10:30:13] src/language-server/block-madness.langium:9:35 - Could not resolve reference to AbstractRule named 'Greeting'.
[10:30:13] src/language-server/block-madness.langium:9:55 - Could not resolve reference to AbstractRule named 'Goodbye'.
[10:30:13] Langium generator failed.
[10:30:18] File change detected. Starting compilation...
[10:30:18] Writing generated files to /home/snarkipus/Projects/langium/langium-race/block-madness/src/language-server/generated
[10:30:18] Writing textmate grammar to /home/snarkipus/Projects/langium/langium-race/block-madness/syntaxes/block-madness.tmLanguage.json
[10:30:18] Langium generator finished successfully in 18ms
```

## The expected behavior
```console
[10:30:18] File change detected. Starting compilation...
[10:30:18] Writing generated files to /home/snarkipus/Projects/langium/langium-race/block-madness/src/language-server/generated
[10:30:18] Writing textmate grammar to /home/snarkipus/Projects/langium/langium-race/block-madness/syntaxes/block-madness.tmLanguage.json
[10:30:18] Langium generator finished successfully in 18ms
```

**Comment by msujew** [comment]
I couldn't reproduce the behavior (on Windows at least). I'm getting a successful CLI run with the provided example:

![image](https://user-images.githubusercontent.com/4377073/214345543-ef1e8991-c842-4c08-abe4-2156a75a61c9.png)

If you want to investigate this yourself, you can use the `Bootstrap` run config of the Langium repo but point the `cwd` towards the `langium-bug` directory. In theory it should hit a breakpoint [here](https://github.com/langium/langium/blob/b5dbb8126c76a87df984a2c39c76974852dbf750/packages/langium/src/references/linker.ts#L211). Maybe you can find the source of the issue from that point onwards.

**Comment by snarkipus** [comment]
> If you want to investigate this yourself, you can use the `Bootstrap` run config of the Langium repo but point the `cwd` towards the `langium-bug` directory ...

I'll take a look - if I find anything interesting, I'll report back.

I was able to re-reproduce it on a windows machine, but I think I missed a step in the 'issue recreation' steps . The initial `watch` generation did actually succeed ... but subsequent executions required the two-step process. To reproduce, I did the following:
 - cloned the repo
 - `npm install`
 - `npm run langium:watch` <-- succeeds 
 - save `super grammar`
 - watch generate fails
 - save 'super grammar`
 - watch generate succeeds

```console
C:\Users\XXXXXXX\Projects\langium-bug-master\langium-bug-master>npm run langium:watch

> block-madness@0.0.1 langium:watch
> langium generate --watch

[12:33:58] Reading config from langium-config.json
[12:33:58] Writing generated files to C:\Users\XXXXXXX\Projects\langium-bug-master\langium-bug-master\src\language-server\generated
[12:33:58] Writing textmate grammar to C:\Users\XXXXXXX\Projects\langium-bug-master\langium-bug-master\syntaxes\block-madness.tmLanguage.json
[12:33:58] Langium generator finished successfully in 270ms
[12:33:58] Langium generator will continue running in watch mode
[12:34:13] File change detected. Starting compilation...
Attempted reference resolution before document reached ComputedScopes state (file:///c%3A/Users/XXXXXXX/Projects/langium-bug-master/langium-bug-master/src/language-server/block-madness-A.langium).
Attempted reference resolution before document reached ComputedScopes state (file:///c%3A/Users/XXXXXXX/Projects/langium-bug-master/langium-bug-master/src/language-server/block-madness-B.langium).
Attempted reference resolution before document reached ComputedScopes state (file:///c%3A/Users/XXXXXXX/Projects/langium-bug-master/langium-bug-master/src/language-server/block-madness-C.langium).
[12:34:13] src\language-server\block-madness.langium:9:15 - Could not resolve reference to AbstractRule named 'Person'.
[12:34:13] src\language-server\block-madness.langium:9:35 - Could not resolve reference to AbstractRule named 'Greeting'.
[12:34:13] src\language-server\block-madness.langium:9:55 - Could not resolve reference to AbstractRule named 'Goodbye'.
[12:34:13] Langium generator failed.
[12:34:18] File change detected. Starting compilation...
[12:34:18] Writing generated files to C:\Users\XXXXXXX\Projects\langium-bug-master\langium-bug-master\src\language-server\generated
[12:34:18] Writing textmate grammar to C:\Users\XXXXXXX\Projects\langium-bug-master\langium-bug-master\syntaxes\block-madness.tmLanguage.json
[12:34:18] Langium generator finished successfully in 56ms
```

**Comment by snarkipus** [comment]
@msujew No joy using the `Bootstrap` version of the `langium-cli`. I suspect it's because it's a one-shot and the generation doesn't actually fail until the second build run.

### launch.json snippet
```json
        {
            "name": "Bootstrap",
            "type": "node",
            "request": "launch",
            "runtimeArgs": [
                "${workspaceFolder}/packages/langium-cli/lib/langium",
                "generate"
            ],
            "runtimeExecutable": "node",
            "cwd": "${workspaceFolder}/examples/langium-bug",
            "skipFiles": [
                "<node_internals>/**"
            ],
            "sourceMaps": true,
            "outFiles": [
                "${workspaceFolder}/packages/langium-cli/lib/**/*.js",
                "${workspaceFolder}/packages/langium/lib/**/*.js"
            ]
        },
```
### Debug Console Output
```console
/home/snarkipus/.nvm/versions/node/v16.19.0/bin/node /home/snarkipus/Projects/langium/langium-new/packages/langium-cli/lib/langium generate
Reading config from langium-config.json
packages/langium-cli/src/generator/util.ts:18
Writing generated files to /home/snarkipus/Projects/langium/langium-new/examples/langium-bug/src/language-server/generated
packages/langium-cli/src/generator/util.ts:18
Writing textmate grammar to /home/snarkipus/Projects/langium/langium-new/examples/langium-bug/syntaxes/block-madness.tmLanguage.json
packages/langium-cli/src/generator/util.ts:18
Langium generator finished successfully in 151ms
```

**Comment by msujew** [comment]
@snarkipus I believe I figured out what breaks the watch script. We made a change in the `invalidateDocument` method, which is used in the generator. It used to actually delete the documents, but now we have a new method for that, `deleteDocument`.

Calling `invalidateDocument` and then trying to re-use these documents led to the warnings you were experiencing. This will be fixed by https://github.com/langium/langium/pull/908.

**Comment by snarkipus** [comment]
@msujew I did try to poke around a bit and assumed it was something related to document lifecycle management. I just couldn't figure out an easy way to debug it ... I ended up in a rabbit hole of complicated test code (to me at least) trying to emulate the scenario. Usually, this means I missed the more obvious way of approaching it ... out of curiosity, what would have been the best way?

**Comment by msujew** [comment]
@snarkipus I was able to reproduce it using a relatively lucky set of circumstances with the `Bootstrap` debug config and adding the `--watch` argument for the `requirements` example project. I was wondering why deleted files weren't really deleted in the documents, which led me to the `invalidateDocument` call. Only then I noticed what we've been doing wrong all along. So in all honesty, it wasn't really obvious, it took me quite a while to figure this out myself.
